src/t_set.c

```diff
@@ -963,18 +1055,20 @@ void sinterGenericCommand(client *c, robj **setkeys,
                 if (limit && cardinality >= limit)
                     break;
             } else if (!dstkey) {
-                if (encoding == OBJ_ENCODING_HT)
-                    addReplyBulkCBuffer(c,elesds,sdslen(elesds));
+                if (str != NULL)
+                    addReplyBulkCBuffer(c, str, len);
                 else
                     addReplyBulkLongLong(c,intobj);
                 cardinality++;
             } else {
-                if (encoding == OBJ_ENCODING_INTSET) {
-                    elesds = sdsfromlonglong(intobj);
+                if (encoding == OBJ_ENCODING_HT) {
+                    sds elesds = (sds)str;
                     setTypeAdd(dstset,elesds);
```

Comment:
let's consider improving the insertion efficiency by improving the above call to createIntsetObject.
we know the cardinality of the dest set in this case.
we can do dictExpand on dict encoded sets, and even keep a track of the total size of elements for pre-setting the capacity of lpNew.

Reply:
don't recall, maybe i was wrong.
maybe we can benefit from setting it to be initially large (according to the size of the first set (which is the smallest), and then scale it down at the end?
i.e. trade repeated rehashing and realloc for one?

Reply:
The dstkey set is created as an intset because if the intersection contains only integers (even if the source sets don't), intset the right encoding for dstkey. We don't have any code for converting to intset from the other encodings.

If we create a listpack of the same capacity as the first set (if the first set is a listpack), then we will not be able to convert it down to intset later (unless we add code for converting listpack to intset).

Reply:
I'll try this and add code for converting to intset. If the first set is a listpack, it means it has some non-integer member and it probably means result is not only integers wither, in most cases, so it can be a good heuristic.

Reply:
we need to keep in mind that insertion into intset can be expensive if done out of order (since its sorted it means insert rather than append), it would end up being beneficial if it remains an intset, but a waste if it is later converted.
on the other hand, if we start with a listpack, we can append out of order, and sort into an intset only if needed.

i'm starting to think that in order to make this efficient, we need quite a lot of heuristics, and code to handle it differently depending on the input sets sizes and encoding.

Reply:
I've pushed a commit attempting some heuristic on this. Please have a look.